sylt-compiler/src/typechecker.rs

```diff
@@ -132,6 +132,151 @@ impl<'c> TypeChecker<'c> {
+                            )
+                        }
+                        known.clone()
+                    }
+                    Entry::Vacant(unknown) => {
+                        unknown.insert(arg.clone()).clone()
+                    }
+                }
+            }
+            (x, y) if x.fits(&y, self.compiler.blobs.as_slice()).is_ok() => x.clone(),
+
+            (Type::Tuple(a), Type::Tuple(b)) => Type::Tuple(a.iter().zip(b.iter()).map(|(a, b)| self.solve_generics_recursively(span, generics, a, b)).collect::<Result<Vec<_>, _>>()?),
+            (Type::List(a), Type::List(b)) => Type::List(Box::new(self.solve_generics_recursively(span, generics, a, b)?)),
+            (Type::Set(a), Type::Set(b)) => Type::Set(Box::new(self.solve_generics_recursively(span, generics, a, b)?)),
+            (Type::Dict(ak, av), Type::Dict(bk, bv)) => Type::Dict(
+                Box::new(self.solve_generics_recursively(span, generics, ak, bk)?),
+                Box::new(self.solve_generics_recursively(span, generics, av, bv)?),
+            ),
+            (Type::Function(a_args, a_ret), Type::Function(b_args, b_ret)) => {
+                let args = a_args.iter().zip(b_args.iter()).map(|(a, b)| self.solve_generics_recursively(span, generics, a, b)).collect::<Result<Vec<_>, _>>()?;
+                let ret = Box::new(self.solve_generics_recursively(span, generics, a_ret, b_ret)?);
+                Type::Function(args, ret)
+            }
+            (a, Type::Union(b)) => {
+                // This is technically wrong, since we might do guesses here that aren't valid.
+                // But it works for our uses.
+                for t in b {
+                    let mut new_generics = generics.clone();
+                    self.solve_generics_recursively(span, &mut new_generics, a, t)?;
+                    *generics = new_generics;
+                }
+                a.clone()
+            }
+            (Type::Union(a), b) => {
+                let ret = a.iter()
+                    .map(|ty| {
+                        let mut new_generics = generics.clone();
+                        let ret = self.solve_generics_recursively(span, &mut new_generics, ty, b);
+                        if ret.is_ok() {
+                            *generics = new_generics;
+                        }
+                        // I feel like there's a better way to do this
+                        // Deliberatly swap, so we get all errors.
+                        match ret {
+                            Ok(x) => Err(x),
+                            Err(x) => Ok(x),
+                        }
+                    })
+                    .collect::<Result<Vec<_>, Type>>();
+                match ret {
+                    Err(ty) => ty,
+                    Ok(errs) => {
+                        return Err(errs.into_iter().fold(Vec::new(), |mut out, mut err| { out.append(&mut err); out }));
+                    }
+                }
+            }
```

Comment:
```suggestion
            (Type::Union(a), b) => {
                a.iter()
                    .map(|ty| {
                        let mut new_generics = generics.clone();
                        let ret = self.solve_generics_recursively(span, &mut new_generics, ty, b);
                        if ret.is_ok() {
                            *generics = new_generics;
                        }
                        ret
                    })
                    // Get the first valid type, or return all errors if no valid type can be found.
                    .fold(Err(Vec::new()), |curr, next| {
                        match (curr, next) {
                            (Ok(ty), _) => Ok(ty),
                            (Err(_), Ok(ty)) => Ok(ty),
                            (Err(mut errs), Err(mut more_errs)) => {
                                errs.append(&mut more_errs);
                                Err(errs)
                            }
                        }
                    })?
            }
```

Reply:
What about this? I think it's clearer what it tries to do. If I understood the original code correctly.

Reply:
But it will iterate over more fields which is why a function would be nice.

Reply:
https://github.com/FredTheDino/sylt-lang/tree/generics-first-ok-or

Reply:
I merge `generics-first-ok-or` - nice